src/tribler/core/components/ipv8/eva_protocol.py

```diff
@@ -144,32 +144,66 @@ class TransferResult:
     data: bytes
     nonce: int
 
+    def __str__(self):
+        return f'TransferResult(peer={self.peer}, info: {self.info}, data hash: {hash(self.data)}, nonce={self.nonce})'
+
+class TransferWindow:
+    def __init__(self, start: int, size: int):
+        self.data: List[Optional[bytes]] = [None for _ in range(size)]
+
+        self.size = size
+        self.start = start
+        self.processed: int = 0
+        self.last_window: bool = False
+
+    def add(self, index: int, block: bytes):
+        if self.data[index] is not None:
+            return
+        self.data[index] = block
+        self.processed += 1
+
+    def finished(self) -> bool:
+        return self.processed == self.size
+
+    def flush(self, data_list: List[bytes]) -> int:
+        added = 0
+        for number in range(self.size):
+            block = self.data[number]
+            if block is None:
+                break
+            data_list.append(block)
+            added += 1
+        return added
+
+    def __str__(self):
+        return f'{{start: {self.start}, processed: {self.processed}, size: {self.size}}}'
+
 
 class Transfer:  # pylint: disable=too-many-instance-attributes
     """The class describes an incoming or an outgoing transfer"""
 
     NONE = -1
 
     def __init__(self, transfer_type: TransferType, info: bytes, data: bytes, data_size: int, block_count: int,
-                 nonce: int, peer: Peer, protocol, future: Optional[Future] = None, window_size: int = 0,
-                 updated: float = 0):
+                 nonce: int, peer: Peer, protocol, future: Optional[Future] = None, updated: float = 0):
         """ This class has been used internally by the EVA protocol"""
         self.type = transfer_type
         self.info = info
         self.data = data
+        self.data_list: List[bytes] = []  # for incoming transfers only
         self.data_size = data_size
         self.block_count = block_count
-        self.block_number = Transfer.NONE
         self.future = future
         self.peer = peer
         self.nonce = nonce
-        self.window_size = window_size
         self.updated = updated
         self.protocol = protocol
 
+        self.window: Optional[TransferWindow] = None
```

Comment:
For incoming and outgoing transfers, the lists of attributes are pretty different, so it looks like they should be presented as two separate subclasses, `IncomingTransfer` and `OutgoingTransfer`.

Reply:
Yes, this is on my TODO list. But I'm not sure that it should be done in this PR.

Reply:
Will be addressed here: https://github.com/Tribler/tribler/pull/6882